File: solutions/dec10/part2.php
<?php
require __DIR__ . "/input.php";

function findNumberOfArrangements($input, $i = 0) {
    var_dump($input);
    $arrangementCount = 0;

    if (!isset($input[$i + 1])) {
        return 1;
    }

    $diff = $input[$i + 1] - $input[$i];
    var_dump($diff);
    if ($diff == 3) {
        findNumberOfArrangements($input, $i+1);
    } elseif ($diff == 2) {
        $index = array_search($input[$i] + 3, $input);
        if ($index !== false) {
            $arrangementCount += findNumberOfArrangements($input, $index);
        }
    } elseif ($diff == 1) {
        if (in_array($input[$i] + 3, $input)) {
            $index = array_search($input[$i] + 3, $input);
            if ($index !== false) {
                $arrangementCount += findNumberOfArrangements($input, $index);
            }
        }

        if (in_array($input[$i] + 2, $input)) {
            $index = array_search($input[$i] + 2, $input);
            if ($index !== false) {
                $arrangementCount += findNumberOfArrangements($input, $index);
            }
        }
    }

    return $arrangementCount;
}

$input = $inputtest2;

$input[] = 0;
$input[] = max($input) + 3;
sort($input);

var_dump(findNumberOfArrangements($input));

File: solutions/dec10/Part2.java
package dec10;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class Part2 {
    private static HashMap<ArrayList<Integer>, Long> resultsCache = new HashMap<>();

    public static void main(String[] args) {
        ArrayList<Integer> input = Input.getInput("input.txt");
        assert input != null;
        input.add(0);
        input.add(Collections.max(input) + 3);
        Collections.sort(input);

        System.out.println(arrangementsCount(input));
    }

    public static long arrangementsCount(ArrayList<Integer> list) {
        if (resultsCache.containsKey(list)) {
            return resultsCache.get(list);
        }

        if (list.size() == 3) {
            return 1;
        }

        long arrangementCount = 1;

        int currentElement;
        int nextElement;

        for (int i = 0; i < list.size() - 1; i++) {
            currentElement = list.get(i);
            nextElement = list.get(i + 1);
            int diff = nextElement - currentElement;

            if (diff <= 2) {
                if (diff != 2) {
                    int indexOfElem2 = list.indexOf(currentElement + 2);
                    if (indexOfElem2 != -1) {
                        arrangementCount += arrangementsCount(new ArrayList<>(list.subList(indexOfElem2, list.size())));
                    }
                }

                int indexOfElem3 = list.indexOf(currentElement + 3);
                if (indexOfElem3 != -1) {
                    arrangementCount += arrangementsCount(new ArrayList<>(list.subList(indexOfElem3, list.size())));
                }
            }
        }

        resultsCache.put(list, arrangementCount);
        return arrangementCount;
    }
}
